hips/hip-9999.md

```diff
@@ -0,0 +1,96 @@
+
+The proposed solution addresses these gaps by integrating a logging library that supports log levels and structured output, as well as a widely supported logger implementation interface for SDK users to consume.
+
+## Rationale
+
+The following libraries were evaluated:
+
+1. **`slog`:**
+   - Provides log levels (`info`, `debug`, `error`, `warning`). More levels can be added
+   - Offers structured logging.
+   - Lightweight and straightforward to integrate.
+   - Is part of the standard library
+   - Has `slog.Handler` interface that other logger implementations can implement to be `slog` compatible
+
+2. **`klog`:**
+   - Provides log levels and also verbosity levels, enabling more granular control. Verbosity level controls the amount of detail in logs, enabling users to filter messages based on their importance or relevance.
+   - Allows getting instrumented Kubernetes client library
+   - Its an external library, hence introducing a dependency
+
+3. **`logr`**
+   - Library defining interfaces for logger implementations and libraries expecting loggers that implement these APIs. `klog`, `slog`, `log` and many others are compatible with `logr`.
+   - Provides log levels also verbosity levels
+   - Its an external library, hence introducing a dependency
+   - Has `logr.Handler` interface that other logger implementations can implement to be `logr` compatible
+
+While these libraries meet all the requirements we want, `logr` is the preferred choice here. It provides all functionality that the other choices have but stand out since it allows using different logger implementations. This will be benefitial to SDK users who have their existing logger implementations present. They would either implement `logr.LogSink` interface or add a dependency of an existing adapter logger implementation.
+
+## Specification
+
+- Helm will instrument `debug`, `info`, `warning` and `error` log levels. These log levels will enable filtering of logs.
+- Logs will be written to `stderr` by the CLI client. `stdout` will be left for output from operations. SDK users would choose where to write logs to through configuring the logger they pass to the SDK.
+- Ability to allow users to configure log levels and formatting through CLI flags or environment variables.
+- Helm SDK should not be instrumented with error logs. Instead, errors ought to be returned. Any logging of such errors should be left to clients to choose whether to log or not.
+- When invoking plugins and post-renderers, `HELM_LOG_LEVEL` environment variable would be set, allowing them to output the appropriate amount of logs to stderr.
+- There will be structured logging with two options. Raw text for humans to view, and JSON formatting that machines can consume.
+- Ability for SDK users to override the default logger implementation. The SDK will expect at logger that implements `logr.Logger` interface.
```

Comment:
This means an application that uses the Helm SDK will need to implement the `logr.Logger` interface with their logger of choice. For example, an application that uses zap will need to create a conversion mechanism to work with the Helm SDK.

The `logr.Logger` interface is not commonly used and I'm aware of numerous applications that use the Helm SDK today but their loggers don't have a conversion mechanism or a native ability to work with `logr.Logger`.

This is what draws me to the `slog` interface. Just about every logging package either natively works with `slog` or has an existing add-on to work with it.

What is the benefit for the SDK to use `logr` and to whom?

Note, the CLI can use logr as an implementation even if the SDK uses slog.

Reply:
> This is what draws me to the slog interface. Just about every logging package either natively works with slog or has an existing add-on to work with it.

This makes sense, plus we won't require clients to depend on a `logr` external package.